### PR TITLE
add color to output

### PR DESCRIPTION
**Pull request by freevryheid**

### CODE REVIEW

**Comment by awvwgk** [comment]
We should add an option to enable/disable the color support. Not all terminals or outputs do support color.

**Comment by ivan-pi** [comment]
I saw I HackerNews thread about https://no-color.org/ recently. They suggest an informal standard of checking the environment variable `NO_COLOR`.

Here's what the [command-line interface guidelines](https://clig.dev/#output) suggest:

    Disable color if your program is not in a terminal or the user requested it. These things should disable colors:

    * stdout or stderr is not an interactive terminal (a TTY). It’s best to individually check—if you’re piping stdout to another program, it’s still useful to get colors on stderr.
    * The NO_COLOR environment variable is set.
    * The TERM environment variable has the value dumb.
    * The user passes the option --no-color.
    * You may also want to add a MYAPP_NO_COLOR environment variable in case users want to disable color specifically for your program.

**Comment by freevryheid** [comment]
I'm leaning towards "the user passes the option -no-color", that flags a public logical no_color defined in testdrive.f90. Since this is a dependency in fpm it is counter-intuitive to include checks that fpm already runs to test for color terminals.

**Comment by ivan-pi** [comment]
I was thinking instead as an optional input to `run_testsuite`. Not sure if I'd want finer grained control at the level of individual test suites, or unit tests, neither at thread/image level.

**Comment by milancurcic** [comment]
I like this PR and agree about having an option to disable colors (env var `NO_COLOR` seems reasonable).

I wonder about the yellow statuses for unexpected pass and expected fail. Personally I'd expect those to be red (undesired outcome) and green (desired outcome), respectively. In a setting where green and red describe desired and undesired outcomes, to me yellow means some kind of warning. And, there's the issue that the same color is used for two opposite outcomes, thus making the coloring unhelpful. Long story short, I suggest not using the yellow, and using green and red according to the desiredness of the test outcome.

**Comment by sloede** [comment]
What's the status of this PR - what is missing for it to get merged? We are using test-drive in our CI testing and unfortunately it is way harder than necessary to visually find problematic locations in the GitHub Actions logs if everything is the same shade of grey...

File: src/testdrive.F90 [hunk 1]
```diff
@@ -553,6 +595,21 @@ function new_testsuite(name, collect) result(self)
...
+
+    !> flag NO_COLOR from environmental variable - run this once when testsuite created
+
+    if (.not.run_once) then
+      call get_environment_variable("NO_COLOR", color_value, STATUS=color_status)
+      if (color_status.eq.0) then
+        NO_COLOR = .true.
+      else
+        NO_COLOR = .false.
+      end if
```
Comment (milancurcic) [inline src/testdrive.F90:609]: Simplify

```suggestion
      NO_COLOR = color_status == 0
```

**Review by milancurcic — APPROVED** [review]

File: src/testdrive.F90 [hunk 2]
```diff
@@ -154,6 +153,34 @@ module testdrive
...
+  character(len=*), parameter :: ESC = char(27)
+
+  !> Escape code for erasing current line
+  ! character(len=*), parameter :: LINE_RESET = ESC//"[2K"//ESC//"[1G"
+
+  !> Escape code for moving up one line
+  ! character(len=*), parameter :: LINE_UP = ESC//"[1A"
+
+  !> Escape code for moving down one line
+  ! character(len=*), parameter :: LINE_DOWN = ESC//"[1B"
```
Comment (milancurcic) [inline src/testdrive.F90:166]: Do we need to keep the commented out code? Likewise for COLOR_YELLOW below.

File: src/testdrive.F90 (hunk 1 above)
Comment (sloede) [reply src/testdrive.F90:609]: LGTM!

File: src/testdrive.F90 (hunk 2 above)
Comment (sloede) [reply src/testdrive.F90:166]: I agree, they should be removed 👍

**Comment by sloede** [comment]
@freevryheid would you be willing to include the proposed changes in your PR such that it can be merged (even though it's been a year...)

**Comment by milancurcic** [comment]
@awvwgk can you confirm that the 3 failing jobs (gnu 6, 7, 8) are due to updated ubuntu-latest image (i.e. older GCC versions not available anymore) and we can safely ignore them?

Also, are you OK with ignoring the codecov coverage decrease in this specific case? It seems to me that adding tests to satisfy the coverage here would be overly tedious for little gain.

File: src/testdrive.F90 [hunk 3]
```diff
@@ -154,6 +153,34 @@ module testdrive
...
+
+  !> Escape code for yellow foreground color
+  ! character(len=*), parameter :: COLOR_YELLOW = ESC//"[93m"
+
+  !> Escape code to reset foreground color
+  character(len=*), parameter :: COLOR_RESET = ESC//"[0m"
+
+  !> color flags
+  logical :: NO_COLOR = .true.
+  logical :: run_once = .false.
```
Comment (awvwgk) [inline src/testdrive.F90:182]: This should be private or a saved variable in the routine it is used

**Comment by awvwgk** [comment]
If we can we should target improving the coverage, but it is not a strict criterium for merging.

Regarding the CI, we should open a new issue on this an fix or remove the respective workflows.

**Pull request closed without merging**